src/transformStateWithClones.js

```diff
@@ -7,7 +7,42 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  // Tablica, która będzie przechowywać historię stanów
+  const stateHistory = [];
+
+  // Zaczynamy od klonowania początkowego stanu
+  let currentState = { ...state };
+
+  // Iterujemy po każdej akcji
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        // Tworzymy nowy pusty obiekt stanu
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        // Dodajemy nowe właściwości do sklonowanego stanu
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        // Klonujemy stan i usuwamy wskazane klucze
+        currentState = { ...currentState };
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
```

Comment:
[CODE STYLE]: It's good practice to use a switch statement when you have a limited amount of conditions, which you have done here. Also, including a default case for error handling is a good way to catch any unexpected action types.